test/e2e/upgrade_management_cluster_test.go

```diff
@@ -2,340 +2,137 @@ package e2e
-		err := downloadFile(clusterctlBinaryPath, clusterctlBinaryURL)
-		Expect(err).ToNot(HaveOccurred(), "failed to download temporary file")
-		defer os.Remove(clusterctlBinaryPath) // clean up
-
-		err = os.Chmod(clusterctlBinaryPath, 0744)
-		Expect(err).ToNot(HaveOccurred(), "failed to chmod temporary file")
-
-		By("Creating a cluster")
-		Logf("Getting the cluster template yaml")
-		upgradeClusterTemplate := clusterctl.ConfigClusterWithBinary(ctx, clusterctlBinaryPath, clusterctl.ConfigClusterInput{
-			KubeconfigPath:       bootstrapClusterProxy.GetKubeconfigPath(),
-			ClusterctlConfigPath: clusterctlConfigPath,
-			// define template variables
-			Namespace:                namespace,
-			ClusterName:              clusterName,
-			KubernetesVersion:        kubernetesVersion,
-			Flavor:                   "ubuntu",
-			ControlPlaneMachineCount: pointer.Int64Ptr(1),
-			WorkerMachineCount:       pointer.Int64Ptr(1),
-			InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
-			// setup clusterctl logs folder
-			LogFolder: clusterctlLogFolder,
-		})
+})
 
-		Expect(upgradeClusterTemplate).ToNot(BeNil(), "Failed to get the cluster template")
+// Install the split-yaml tool to split our bmhosts_crs.yaml file.
+func installSplitYaml() {
+	cmd := exec.Command("bash", "-c", "kubectl krew update; kubectl krew install split-yaml")
+	output, err := cmd.CombinedOutput()
+	Logf("Download split-yaml:\n %v", string(output))
+	Expect(err).To(BeNil())
+}
 
-		Logf("Applying the cluster template yaml to the cluster")
+// Split the bmhosts_crs.yaml file to make it easy to apply limited number on BMH in each namespace.
+func splitBMHostFile() {
+	cmd := exec.Command("bash", "-c", "cat bmhosts_crs.yaml | kubectl split-yaml -t {{.name}}.yaml -p.")
+	cmd.Dir = workDir
+	output, err := cmd.CombinedOutput()
+	Logf("splitting bmhs:n %v", string(output))
+	Expect(err).To(BeNil())
+}
 
-		Expect(bootstrapClusterProxy.Apply(ctx, upgradeClusterTemplate)).To(Succeed())
+func getPreWaitForCluster(ctx context.Context) func(proxy framework.ClusterProxy, clusterNamespace string, clusterName string) {
+	return func(clusterProxy framework.ClusterProxy, clusterNamespace string, clusterName string) {
+		installSplitYaml()
+		splitBMHostFile()
+
+		// Apply bmh and bmc-secret file based on node index
+		applyBMH := func(node_index int) {
+			node := fmt.Sprintf("node-%v", node_index)
+			cmd := exec.Command("kubectl", "apply", "-f", fmt.Sprintf("/opt/metal3-dev-env/%v-bmc-secret.yaml", node), "-n", clusterNamespace, "--kubeconfig", clusterProxy.GetKubeconfigPath())
+			cmd.Dir = workDir
+			output, err := cmd.CombinedOutput()
+			Logf(string(output))
+			Expect(err).To(BeNil())
+			cmd = exec.Command("kubectl", "apply", "-f", fmt.Sprintf("/opt/metal3-dev-env/%v.yaml", node), "-n", clusterNamespace, "--kubeconfig", clusterProxy.GetKubeconfigPath())
+			cmd.Dir = workDir
+			output, err = cmd.CombinedOutput()
```

Comment:
If we change the `splitBMHostFile` like suggested, we could make this easier to understand by looping over the file names directly. I would also want to use `clusterProxy.Apply` like for cert-manager below.

Reply:
done